File: lc_java/lcJava/greedy/MaxAverageRatio.java
import java.util.*;

class Solution {
    public double maxAverageRatio(int[][] classes, int extraStudents) {
        // 使用优先队列，按增益降序排列
        PriorityQueue<double[]> pq = new PriorityQueue<>((a, b) -> 
            Double.compare(b[0], a[0])
        );
        
        // 初始化堆
        for (int[] c : classes) {
            double pass = c[0], total = c[1];
            double gain = (pass + 1) / (total + 1) - pass / total;
            pq.offer(new double[]{gain, pass, total});
        }
        
        // 分配额外学生
        for (int i = 0; i < extraStudents; i++) {
            double[] top = pq.poll();
            double pass = top[1] + 1;
            double total = top[2] + 1;
            double gain = (pass + 1) / (total + 1) - pass / total;
            pq.offer(new double[]{gain, pass, total});
        }
        
        // 计算平均通过率
        double totalRatio = 0.0;
        while (!pq.isEmpty()) {
            double[] top = pq.poll();
            totalRatio += top[1] / top[2];
        }
        
        return totalRatio / classes.length;
    }
    
    public static void main(String[] args) {
        Solution s = new Solution();
        
        // 测试用例1
        int[][] classes1 = {{1,2},{3,5},{2,2}};
        int extraStudents1 = 2;
        double result1 = s.maxAverageRatio(classes1, extraStudents1);
        System.out.printf("测试用例1: %.5f%n", result1);
        
        // 测试用例2
        int[][] classes2 = {{2,4},{3,9},{4,5},{2,10}};
        int extraStudents2 = 4;
        double result2 = s.maxAverageRatio(classes2, extraStudents2);
        System.out.printf("测试用例2: %.5f%n", result2);
    }
}

File: lc_go/lcGo/greedy/maxAverageRatio.go
package main

import (
    "container/heap"
    "fmt"
)

type ClassInfo struct {
    gain  float64
    pass  int
    total int
}

type MaxHeap []ClassInfo

func (h MaxHeap) Len() int           { return len(h) }
func (h MaxHeap) Less(i, j int) bool { return h[i].gain > h[j].gain }
func (h MaxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *MaxHeap) Push(x interface{}) {
    *h = append(*h, x.(ClassInfo))
}

func (h *MaxHeap) Pop() interface{} {
    old := *h
    n := len(old)
    x := old[n-1]
    *h = old[0 : n-1]
    return x
}

func calculateGain(pass, total int) float64 {
    return float64(pass+1)/float64(total+1) - float64(pass)/float64(total)
}

func maxAverageRatio(classes [][]int, extraStudents int) float64 {
    h := &MaxHeap{}
    heap.Init(h)
    
    // 初始化堆
    for _, class := range classes {
        pass, total := class[0], class[1]
        gain := calculateGain(pass, total)
        heap.Push(h, ClassInfo{gain, pass, total})
    }
    
    // 分配额外学生
    for i := 0; i < extraStudents; i++ {
        top := heap.Pop(h).(ClassInfo)
        pass := top.pass + 1
        total := top.total + 1
        gain := calculateGain(pass, total)
        heap.Push(h, ClassInfo{gain, pass, total})
    }
    
    // 计算平均通过率
    totalRatio := 0.0
    for h.Len() > 0 {
        top := heap.Pop(h).(ClassInfo)
        totalRatio += float64(top.pass) / float64(top.total)
    }
    
    return totalRatio / float64(len(classes))
}

func main() {
    // 测试用例1
    classes1 := [][]int{{1, 2}, {3, 5}, {2, 2}}
    extraStudents1 := 2
    result1 := maxAverageRatio(classes1, extraStudents1)
    fmt.Printf("测试用例1: %.5f\n", result1)
    
    // 测试用例2
    classes2 := [][]int{{2, 4}, {3, 9}, {4, 5}, {2, 10}}
    extraStudents2 := 4
    result2 := maxAverageRatio(classes2, extraStudents2)
    fmt.Printf("测试用例2: %.5f\n", result2)
}